pkg/plugins/placementpolicy/core/policy.go

```diff
@@ -0,0 +1,233 @@
+	}
+
+	created := newPolicyInfo(policyNamespace, policyName, action, targetSize)
+	pi.internal[policyNamespace][policyName] = created
+
+	pi.Unlock()
+
+	return created
+}
+
+// RemovePolicy removes the policy from the in-memory collection
+func (pi *PolicyInfos) RemovePolicy(policyNamespace string, policyName string) {
+	pi.Lock()
+	delete(pi.internal[policyNamespace], policyName)
+	pi.Unlock()
+}
+
+// GetPoliciesByNamespace returns all policies in the namespace
+func (pi *PolicyInfos) GetPoliciesByNamespace(policyNamespace string) map[string]*PolicyInfo {
+	pi.RLock()
+	namespacePolicies := pi.internal[policyNamespace]
+	pi.RUnlock()
+	return namespacePolicies
+}
+
+// UpdatePolicy updates the policy held in-memory
+func (pi *PolicyInfos) UpdatePolicy(policyNamespace string, policy *PolicyInfo) {
+	policyName := policy.Name
+	pi.Lock()
+	existing := pi.internal[policyNamespace][policyName]
+	pi.internal[policyNamespace][policyName] = policy.merge(existing)
+	pi.Unlock()
+}
+
+func (p *PolicyInfo) merge(existing *PolicyInfo) *PolicyInfo {
+	existing.Action = p.Action
+	existing.TargetSize = p.TargetSize
+	existing.targetMet = p.targetMet
+
+	tempMatched := sets.NewString()
+	if len(p.matchedPods) > 0 {
+		tempMatched = tempMatched.Insert(p.matchedPods.List()...)
+	}
+	existing.matchedPods = tempMatched
+
+	tempQualified := sets.NewString()
+	if len(p.qualifiedPods) > 0 {
+		tempQualified = tempQualified.Insert(p.qualifiedPods.List()...)
+	}
+	existing.qualifiedPods = tempQualified
+
+	tempManaged := sets.NewString()
+	if len(p.managedPods) > 0 {
+		tempManaged = tempManaged.Insert(p.managedPods.List()...)
+	}
+	existing.managedPods = tempManaged
+
+	return existing
+}
+
+// addMatch adds the pod to the policy's matchedPods collection
+func (p *PolicyInfo) addMatch(pod *corev1.Pod) error {
+	key, err := framework.GetPodKey(pod)
+	if err != nil {
+		return err
+	}
+
+	p.matchedPods = p.matchedPods.Insert(key)
+	return nil
+}
+
+// removePodIfPresent removes the pod from the policy's qualifiedPods and managedPods collections as appropriate
+func (p *PolicyInfo) removePodIfPresent(pod *corev1.Pod) error {
+	key, err := framework.GetPodKey(pod)
+	if err != nil {
+		return err
+	}
+
+	if !p.PodQualifiesForPolicy(key) {
+		return nil
+	}
+
+	p.qualifiedPods = p.qualifiedPods.Delete(key)
+
+	if p.PodIsManagedByPolicy(key) {
+		p.managedPods = p.managedPods.Delete(key)
+	}
+
+	return p.setTargetMet()
+}
+
+// addPodIfNotPresent adds the pod to the policy's qualifiedPods and managedPods collections as appropriate
+func (p *PolicyInfo) addPodIfNotPresent(pod *corev1.Pod) error {
+	key, keyError := framework.GetPodKey(pod)
+	if keyError != nil {
+		return keyError
+	}
+
+	//if pod is already in the list, do nothing
+	if p.PodQualifiesForPolicy(key) {
+		return nil
+	}
+
+	//if policy is `BestEffort`, this will be true
+	if p.PodMatchesPolicy(key) {
+		p.matchedPods = p.matchedPods.Delete(key) //once added, don't need to worry about matched anymore
```

Comment:
nit: move the comment to the line above
```suggestion
                // once added, don't need to worry about matched anymore
		p.matchedPods = p.matchedPods.Delete(key) 
```